### PR TITLE
chore: unify and optimize Rust CI

### LINKED ISSUE #1714: Rust workspace no longer works on Linux
I can no longer compile the Rust workspace on my machine because I am hitting the following panic:

https://github.com/firezone/firezone/blob/1882b58bc4f9b98895f62757f1ad20ed0cb5b713/rust/connlib/clients/apple/build.rs#L59

Is there a way we can define this build-script to gracefully fail on non-Mac machines? My workflow is to typically run `cargo clippy --all-targets --all-features` during development and also my IDE indexes all crates in the workspace.

For both of these, it would be great if we can have the build-script not panic.

### PR DESCRIPTION
**Pull request by thomaseizinger**
- Instead of having two, very similar jobs, we run our fmt, clippy and tests steps across all crates and operating systems.
- We remove the dependency of the android and apple builds on the tests and thus get faster feedback.
- We force clippy to fail on any warning. This one is super important IMO. Warnings in Rust are very useful and ignoring them can lead to bugs (think "unused Result" etc).

Resolves #1714.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
https://github.com/firezone/firezone/pull/1691 fixes the android and apple builds. I'll wait for that one to be merged and then make minor adjustments here for the ring build etc.

**Comment by thomaseizinger** [comment]
@francesca64 Do you have an idea why the apple build is failing in here? Help would be much appreciated :)

**Comment by francesca64** [comment]
@thomaseizinger build scripts themselves are compiled for the host machine, so the conditional compilation you added to `build.rs` doesn't do what you'd expect it to when cross-compiling. That said, on CI we're only cross-compiling from macOS -> more Apple, so the error is still unexpected. I'll see if I can repro.

File: rust/connlib/clients/apple/build.rs
Comment (conectado) [inline rust/connlib/clients/apple/build.rs]:
```suggestion
// TODO: rust-lang/cargo#5220
#[cfg(any(target_os = "darwin", target_os = "ios", target_os = "ios-sim"))]
```
I think when that is closed we should just do platform-specific workspaces.


Also, @thomaseizinger I think the build is failing because you're using `target_os = "darwin"` instead of `target_os = "macos"` ([see available `target_os` options](https://doc.rust-lang.org/reference/conditional-compilation.html#target_os))

Comment (conectado) [reply rust/connlib/clients/apple/build.rs]: cc @francesca64

File: .github/workflows/rust.yml
Comment (jamilbk) [inline .github/workflows/rust.yml:97]: 🙏

File: rust/connlib/clients/apple/build.rs
Comment (jamilbk) [inline rust/connlib/clients/apple/build.rs]: Weird that Rust uses `darwin` in the target triple but `macos` for `target_os`
```suggestion
#[cfg(not(any(target_os = "macos", target_os = "ios", target_os = "ios-sim")))]
```

**Review by jamilbk — CHANGES_REQUESTED** [review]
Found a minor issue

**Comment by thomaseizinger** [comment]
> @thomaseizinger build scripts themselves are compiled for the host machine, so the conditional compilation you added to `build.rs` doesn't do what you'd expect it to when cross-compiling.

I expected the build script to do nothing on non-MacOS machines because those don't have XCode. I think that is what I achieved but I am open to other solutions :)

The apple-crates themselves aren't going to work but I think as long as we don't try to link them on other operating systems, that should be fine?

**Comment by thomaseizinger** [comment]
@firezone/engineering The relay smoke test script seems to be a bit flaky. I am looking into it. For this PR, I just restarted the job now.

**Comment by francesca64** [comment]
@thomaseizinger 
> I expected the build script to do nothing on non-MacOS machines because those don't have XCode. I think that is what I achieved but I am open to other solutions :)

It's a fair point that cross-compiling wouldn't work anyway, but the correct solution here will be cleaner regardless. Checking `CARGO_CFG_TARGET_VENDOR` at the top of `main` and then early-returning if it's not `apple` would reduce the `build.rs` diff to only 3 lines, and it's also the pattern we need to remember to follow for build scripts in general.

File: rust/connlib/clients/apple/build.rs
Comment (francesca64) [inline rust/connlib/clients/apple/build.rs]:
```suggestion
    if std::env::var("CARGO_CFG_TARGET_VENDOR").as_deref() != Ok("apple") {
```
`cfg!` has the same caveat as `cfg`, i.e. if you cross-compiled macOS -> Linux then `build.rs` would erroneously run.

**Review by francesca64 — CHANGES_REQUESTED** [review]

**Review by francesca64 — APPROVED** [review]

**Review by jamilbk — APPROVED** [review]

**Comment by jamilbk** [comment]
Waiting on @conectado to approve. Should code owners be updated so there's redundancy to prevent blocking?

**Review by conectado — APPROVED** [review]

**Pull request merged**
